README.md
# Hanami::Helpers

View helpers for Ruby applications

## Status

[![Gem Version](https://badge.fury.io/rb/hanami-helpers.svg)](https://badge.fury.io/rb/hanami-helpers)
[![CI](https://github.com/hanami/helpers/workflows/ci/badge.svg?branch=main)](https://github.com/hanami/helpers/actions?query=workflow%3Aci+branch%3Amain)
[![Test Coverage](https://codecov.io/gh/hanami/helpers/branch/main/graph/badge.svg)](https://codecov.io/gh/hanami/helpers)
[![Depfu](https://badges.depfu.com/badges/7b30c5d2a2a78954e2ad86e5c7230c23/overview.svg)](https://depfu.com/github/hanami/helpers?project=Bundler)
[![Inline Docs](http://inch-ci.org/github/hanami/helpers.svg)](http://inch-ci.org/github/hanami/helpers)

## Contact

* Home page: http://hanamirb.org
* Mailing List: http://hanamirb.org/mailing-list
* API Doc: http://rdoc.info/gems/hanami-helpers
* Bugs/Issues: https://github.com/hanami/helpers/issues
* Support: http://stackoverflow.com/questions/tagged/hanami
* Chat: https//chat.hanamirb.org

## Rubies

__Hanami::Helpers__ supports Ruby (MRI) 3.0+

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'hanami-helpers'
```

And then execute:

```shell
$ bundle
```

Or install it yourself as:

```shell
$ gem install hanami-helpers
```

## Usage

`Hanami::Helpers` offers a set of utilities to enrich web views.

### HTML helper

HTML5 markup generator (`#html`).

View:

```ruby
module Users
  class Show
    include Hanami::Helpers

    def sidebar
      html.aside(id: 'sidebar') do
        p "Languages", class: 'title'

        ul do
          li "Italian"
          li "English"
        end
      end
    end
  end
end
```

Template:

```erb
<%= sidebar %>
```

Output:

```html
<aside id="sidebar">
  <p class="title">Languages</p>

  <ul>
    <li>Italian</li>
    <li>English</li>
  </ul>
</aside>
```

### Form Helper

Form generator for HTML5 (`#form_for`)

#### Template Usage

Template:

```erb
<%=
  form_for :book, routes.books_path do
    text_field :title

    submit 'Create'
  end
%>
```

Output:

```html
<form action="/books" method="POST" accept-charset="utf-8" id="book-form">
  <input type="text" name="book[title]" id="book-id" value="">
  <button type="submit">Create</button>
</form>
```

#### View Usage

View:

```ruby
module Books
  class New
    include Hanami::Helpers

    def form
      form_for :book, routes.books_path do
        text_field :title

        submit 'Create'
      end
    end
  end
end
```

Template:

```erb
<%= form %>
```

Output:

```html
<form action="/books" method="POST" accept-charset="utf-8" id="book-form">
  <input type="text" name="book[title]" id="book-id" value="">
  <button type="submit">Create</button>
</form>
```

#### Reuse Code

Views:

```ruby
module Books
  class New
    include Hanami::Helpers

    def form
      Form.new(:book, routes.books_path)
    end

    def submit_label
      'Create'
    end
  end

  class Edit
    include Hanami::Helpers

    def form
      Form.new(:book, routes.book_path(id: book.id), {book: book}, {method: :patch})
    end

    def submit_label
      'Update'
    end
  end
end
```

Templates:

```erb
# books/new.html.erb
<%= render partial: 'books/form' %>
```

```erb
# books/edit.html.erb
<%= render partial: 'books/form' %>
```

```erb
# books/_form.html.erb
<%=
  form_for form, class: 'form-horizontal' do
    text_field :title

    submit submit_label
  end
%>
```

Output for new:

```html
<form action="/books" method="POST" accept-charset="utf-8" id="book-form">
  <input type="text" name="book[title]" id="book-id" value="">
  <button type="submit">Create</button>
</form>
```

Output for edit:

```html
<form action="/books/23" method="POST" accept-charset="utf-8" id="book-form">
  <input type="hidden" name="_method" value="PATCH">
  <input type="text" name="book[title]" id="book-id" value="TDD">
  <button type="submit">Update</button>
</form>
```

### Escape helper

HTML (`#h`), HTML attribute (`#ha`) and URL (`#hu`) escape helpers.

View:

```ruby
module Users
  class Show
    include Hanami::Helpers

    def home_page_link
      %(<a href="#{ hu(user.home_page_url) }" title="#{ ha(user.name} }'s website">#{ h(user.website_name) }</a>)
    end

    def code_snippet
      raw user.code_snippet
    end
  end
end
```

Template:

```erb
<%= home_page_link %>
<%= code_snippet %>
```

Output:

```html
<a href="https://example.org" title="Maria's website">My Blog</a>
<code>puts "Hello, World!"</code>
```

### Routing Helper

Hanami and Hanami::Router integration (`#routes`).

View:

```ruby
module Home
  class Index
    include Hanami::Helpers

    def link_to_home
      %(<a href="#{ routes.home_path }">Home</a>)
    end
  end
end
```

Template:

```erb
<%= link_to_home %>
```

Output:

```html
<a href="/">Home</a>
```

### Number Formatting Helper

Format numbers (`#format_number`).

View:

```ruby
module Home
  class Index
    include Hanami::Helpers

    def visitors_count
      format_number '1000'
    end
  end
end
```

Template:

```erb
<p><%= visitors_count %></p>
```

Output:

```html
<p>1,000</p>
```

## Philosophy

All the Hanami helpers are modules to include.

Most of the time they inject **private** methods.
This restriction prevents helper methods to be used on the outside (eg. in a template).

We want to encourage developers to use **meaningful** and **simple APIs** in their templates.

### Bad style example

```ruby
module Users
  class Show
    include Hanami::Helpers
  end
end
```

```erb
<%= format_number user.followers_count %>
```

This style increases the complexity of the template and it makes testing hard.

### Good style example

```ruby
module Users
  class Show
    include Hanami::Helpers

    def followers_count
      format_number user.followers_count
    end
  end
end
```

```erb
<%= followers_count %>
```

This simplifies the markup.
In order to test the value that will be printed becomes easier: `Users::Show#followers_count`.

## Versioning

__Hanami::Helpers__ uses [Semantic Versioning 2.0.0](http://semver.org)

## Contributing

1. Fork it ( https://github.com/hanami/helpers/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request

## Copyright

Copyright © 2014-2021 Luca Guidi – Released under MIT License

This project was formerly known as Lotus (`lotus-helpers`).


CHANGELOG.md
# Hanami::Helpers
View helpers for Ruby web applications

## v2.0.0.alpha8 (unreleased)
### Fixed
- [Luca Guidi] `Hanami::Helpers::NumberFormattingHelper#format_number` to return consistent rounding. `1.0` will be formatted as `"1.00"` by default.

### Changed
- [Luca Guidi] Drop support for Ruby: MRI 2.3, 2.4, 2.5, 2.6, and 2.7
- [Luca Guidi] Moved `Hanami::Helpers::RoutingHelper` to `hanami` as `Hanami::Helpers::Routes`
- [Luca Guidi] Made `Hanami::Helpers::NumberFormattingHelper#format_number` a public method
- [Luca Guidi] `Hanami::Helpers::NumberFormattingHelper#format_number` raises a `Hanami::Helpers::CoercionError` instead of `TypeError`
- [Luca Guidi] `Hanami::Helpers::NumberFormattingHelper#form_for` removed the first argument that used to indicate the form name prefix
    ```ruby
    # 1.0
    form_for(:book, routes.books_path) {}

    #   => <form action="/path" method="POST" id="book-form"></form>

    # 2.0
    form_for(routes.books_path, id: "book-form") {}

    #   => <form action="/path" method="POST" id="book-form"></form>
    ```
- [Luca Guidi] `Hanami::Helpers::NumberFormattingHelper#form_for` yields a `Hanami::Helpers::FormHelper::FormBuilder` object that MUST be used to generate form inputs
    ```ruby
    # 1.0
    form_for(:book, routes.books_path) do
      text_field :title

      submit "Create book"
    end

    # <form action="/path" method="POST" id="book-form">
    #   <input type="text" name="book[title]" id="book-title">
    #   <button type="submit">Create book</button>
    # </form>

    # 2.0
    form_for(routes.books_path, id: "book-form") do |f|
      f.text_field "book.title"

      f.submit "Create book"
    end

    # <form action="/path" method="POST" id="book-form">
    #   <input type="text" name="book[title]" id="book-title">
    #   <button type="submit">Create book</button>
    # </form>
    ```
- [Luca Guidi] Form input helpers now require full path to be always specified as a dot separated `String`
    ```ruby
    # 1.0
    form_for(:book, routes.books_path) do
      fields_for :author do
        fields_for :avatar do
          text_field :url
        end
      end

      submit "Create book"
    end

    # <form action="/path" method="POST" id="book-form">
    #   <input type="text" name="book[author][avatar][url]" id="book-author-avatar-url">
    #   <button type="submit">Create book</button>
    # </form>

    # 2.0
    form_for(routes.books_path, id: "book-form") do |f|
      f.text_field "book.author.avatar.url"

      f.submit "Create book"
    end

    # <form action="/path" method="POST" id="book-form">
    #   <input type="text" name="book[author][avatar][url]" id="book-author-avatar-url">
    #   <button type="submit">Create book</button>
    # </form>
    ```
- [Luca Guidi] Removed `fields_for` form helper

## v1.3.3 - 2020-02-03
### Added
- [Luca Guidi] Official support for Ruby: MRI 2.7

## v1.3.2 - 2019-10-31
### Added
- [Nenad] Support block syntax for `label` helper
- [Luca Guidi] Allow `form_for` to accept `params:` argument to override params coming from the view

### Fixed
- [Danny Santos] When `prompt:` option is used in `select` helper, set to `disabled="disabled"` the related `<option>` HTML tag
>>>>>>> develop

## v1.3.1 - 2019-01-18
### Added
- [Luca Guidi] Official support for Ruby: MRI 2.6
- [Luca Guidi] Support `bundler` 2.0+

## v1.3.0 - 2018-10-24

## v1.3.0.beta1 - 2018-08-08
### Added
- [Luca Guidi] Official support for JRuby 9.2.0.0

## v1.2.2 - 2018-06-12
### Fixed
- [Sean Collins] Ensure `csrf_meta_tags` helper to generate `<meta>` tags with `content` attribute instead of `value`

## v1.2.1 - 2018-06-04
### Fixed
- [Lucas Gomes] Ensure to mark `<option>` as selected when `select` helper has a collection of non-`String` values

## v1.2.0 - 2018-04-11

## v1.2.0.rc2 - 2018-04-06
### Added
- [Ponomarev Ilya] Allow `submit` and `button` form helpers to accept blocks
- [Ferdinand Niedermann] let `fields_for_collection` to iterate thru the given collection and yield current index and value

## v1.2.0.rc1 - 2018-03-30
### Fixed
- [Gustavo Caso] Ensure `select` helper to set the `selected` attribute properly when an `<option>` has a `nil` value

## v1.2.0.beta2 - 2018-03-23

## v1.2.0.beta1 - 2018-02-28
### Added
- [Luca Guidi] Introduced `csrf_meta_tags` helper to print meta tags for CSRF protection
- [Luca Guidi] Added support for `form_for(..., remote: true)` which translates to `<form ... data-remote="true">`
- [Sean Collins] Allow HTML attributes to be specified as an array of strings or symbols (eg `<%= html.span("foo", class: %w(ui form)) %>`)

## v1.1.2 - 2018-04-05
### Fixed
- [Luca Guidi] Ensure correct arity of `#form_for`, to be used in conjuction with `Hanami::Helpers::FormHelper::Form` (eg. `<%= form_for(form) { ... } %>`)

## v1.1.1 - 2018-02-27
### Fixed
- [Alfonso Uceda] Ensure `#select` form helper to not select options with `nil` value
- [Alfonso Uceda] Ensure `#fields_for_collection` form helper to produce input fields with correct `name` attribute
- [Luca Guidi] Ensure `#select` form helper to respect `:selected` option

## v1.1.0 - 2017-10-25

## v1.1.0.rc1 - 2017-10-16

## v1.1.0.beta3 - 2017-10-04

## v1.1.0.beta2 - 2017-10-03

## v1.1.0.beta1 - 2017-08-11

## v1.0.0 - 2017-04-06

## v1.0.0.rc1 - 2017-03-31

## v1.0.0.beta2 - 2017-03-17
### Added
- [Luca Guidi] Added `time_field` form helper
- [Luca Guidi] Added `month_field` form helper
- [Luca Guidi] Added `week_field` form helper
- [Luca Guidi] Added `range_field` form helper
- [Luca Guidi] Added `search_field` form helper
- [Luca Guidi] Added `url_field` form helper
- [Luca Guidi] Added `tel_field` form helper
- [Luca Guidi] Added `image_button` form helper
- [Luca Guidi] Added support for `<dialog>`, `<hgroup>`, `<rtc>`, `<slot>`, and `<var>` HTML5 tags

## v1.0.0.beta1 - 2017-02-14
### Added
- [Luca Guidi] Official support for Ruby: MRI 2.4

lib/hanami/helpers/form_helper.rb
# frozen_string_literal: true

require "hanami/helpers/form_helper/form_builder"
require "hanami/helpers/html_helper"

module Hanami
  module Helpers
    # Form builder
    #
    # By including <tt>Hanami::Helpers::FormHelper</tt> it will inject one public method: <tt>form_for</tt>.
    # This is a HTML5 form builder.
    #
    # To understand the general HTML5 builder syntax of this framework, please
    # consider to have a look at <tt>Hanami::Helpers::HtmlHelper</tt> documentation.
    #
    # This builder is independent from any template engine.
    # This was hard to achieve without a compromise: the form helper should be
    # used in one output block in a template or as a method in a view (see the examples below).
    #
    # Features:
    #
    #   * Support for complex markup without the need of concatenation
    #   * Auto closing HTML5 tags
    #   * Support for view local variables
    #   * Method override support (PUT/PATCH/DELETE HTTP verbs aren't understood by browsers)
    #   * Automatic generation of HTML attributes for inputs: <tt>id</tt>, <tt>name</tt>, <tt>value</tt>
    #   * Allow to override HTML attributes
    #   * Extract values from request params and fill <tt>value</tt> attributes
    #   * Automatic selection of current value for radio button and select inputs
    #   * Infinite nested fields
    #
    # Supported tags and inputs:
    #
    #   * <tt>check_box</tt>
    #   * <tt>color_field</tt>
    #   * <tt>date_field</tt>
    #   * <tt>datetime_field</tt>
    #   * <tt>datetime_local_field</tt>
    #   * <tt>email_field</tt>
    #   * <tt>fields_for</tt>
    #   * <tt>file_field</tt>
    #   * <tt>form_for</tt>
    #   * <tt>hidden_field</tt>
    #   * <tt>label</tt>
    #   * <tt>number_field</tt>
    #   * <tt>password_field</tt>
    #   * <tt>radio_button</tt>
    #   * <tt>select</tt>
    #   * <tt>submit</tt>
    #   * <tt>text_area</tt>
    #   * <tt>text_field</tt>
    #
    # @since 0.2.0
    #
    # @see Hanami::Helpers::FormHelper#form_for
    # @see Hanami::Helpers::HtmlHelper
    #
    # @example One output block (template)
    #   <%=
    #     form_for :book, routes.books_path do
    #       text_field :title
    #
    #       submit 'Create'
    #     end
    #   %>
    #
    # @example Method (view)
    #   require 'hanami/helpers'
    #
    #   class MyView
    #     include Hanami::Helpers::FormHelper
    #
    #     def my_form
    #       form_for :book, routes.books_path do
    #         text_field :title
    #       end
    #     end
    #   end
    #
    #   <!-- use this in the template -->
    #   <%= my_form %>
    module FormHelper
      # Default HTTP method for form
      #
      # @since 0.2.0
      # @api private
      DEFAULT_METHOD = "POST"

      # Default charset
      #
      # @since 0.2.0
      # @api private
      DEFAULT_CHARSET = "utf-8"

      # CSRF Token session key
      #
      # This key is shared with <tt>hanamirb</tt>, <tt>hanami-controller</tt>.
      #
      # @since 0.2.0
      # @api private
      CSRF_TOKEN = :_csrf_token

      include HtmlHelper

      # Form object
      #
      # @since 0.2.0
      class Form
        # @return [Symbol] the form name
        #
        # @since 0.2.0
        # @api private
        attr_reader :name

        # @return [String] the form action
        #
        # @since 0.2.0
        # @api private
        attr_reader :url

        # @return [::Hash] the form values
        #
        # @since 0.2.0
        # @api private
        attr_reader :values

        # Initialize a form
        #
        # It accepts a set of values that are used in combination with request
        # params to autofill <tt>value</tt> attributes for fields.
        #
        # The keys of this Hash, MUST correspond to the structure of the (nested)
        # fields of the form.
        #
        # For a given input where the <tt>name</tt> is `book[title]`, Hanami will
        # look for `:book` key in values.
        #
        # If the current params have the same key, it will be PREFERRED over the
        # given values.
        #
        # For instance, if <tt>params.get('book.title')</tt> equals to
        # <tt>"TDD"</tt> while <tt>values[:book].title</tt> returns
        # <tt>"No test"</tt>, the first will win.
        #
        # @param name [Symbol] the name of the form
        # @param url [String] the action of the form
        # @param values [Hash,NilClass] a Hash of values to be used to autofill
        #   <tt>value</tt> attributes for fields
        # @param attributes [Hash,NilClass] a Hash of attributes to pass to the
        #   <tt>form</tt> tag
        #
        # @since 0.2.0
        #
        # @example Pass A Value
        #   # Given the following view
        #
        #   module Web::Views::Deliveries
        #     class Edit
        #       include Web::View
        #
        #       def form
        #         Form.new(:delivery, routes.delivery_path(id: delivery.id),
        #         {delivery: delivery, customer: customer},
        #         {method: :patch})
        #       end
        #     end
        #   end
        #
        #   # And the corresponding template:
        #
        #   <%=
        #     form_for form do
        #       date_field :delivered_on
        #
        #       fields_for :customer do
        #         text_field :name
        #
        #         fields_for :address do
        #           # ...
        #           text_field :city
        #         end
        #       end
        #
        #       submit 'Update'
        #     end
        #   %>
        #
        #   <!-- output -->
        #
        #   <form action="/deliveries/1" method="POST" accept-charset="utf-8">
        #     <input type="hidden" name="_method" value="PATCH">
        #     <input type="hidden" name="_csrf_token" value="920cd5bfaecc6e58368950e790f2f7b4e5561eeeab230aa1b7de1b1f40ea7d5d">
        #
        #     # Value taken from delivery.delivered_on
        #     <input type="date" name="delivery[delivered_on]" id="delivery-delivered-on" value="2015-05-27">
        #
        #     # Value taken from customer.name
        #     <input type="text" name="delivery[customer][name]" id="delivery-customer-name" value="Luca">
        #
        #     # Value taken from customer.address.city
        #     <input type="text" name="delivery[customer][address][city]" id="delivery-customer-address-city" value="Rome">
        #
        #     <button type="submit">Update</button>
        #   </form>
        def initialize(name, url, values = {}, attributes = {})
          @name       = name
          @url        = url
          @values     = values
          @attributes = attributes || {}
        end

        # Return the method specified by the given attributes or fall back to
        # the default value
        #
        # @return [String] the method for the action
        #
        # @since 0.2.0
        # @api private
        #
        # @see Hanami::Helpers::FormHelper::DEFAULT_METHOD
        def verb
          @attributes.fetch(:method, DEFAULT_METHOD)
        end
      end

      # Instantiate a HTML5 form builder
      #
      # @overload form_for(name, url, options, &blk)
      #   Use inline values
      #   @param name [Symbol] the toplevel name of the form, it's used to generate
      #     input names, ids, and to lookup params to fill values.
      #   @param url [String] the form action URL
      #   @param options [Hash] HTML attributes to pass to the form tag and form values
      #   @option options [Hash] :values An optional payload of objects to pass
      #   @option options [Hash] :params An optional override of params
      #   @param blk [Proc] A block that describes the contents of the form
      #
      # @overload form_for(form, attributes = {}, &blk)
      #   Use Form
      #   @param form [Hanami::Helpers::FormHelper::Form] a form object
      #   @param attributes [Hash] HTML attributes to pass to the form tag and form values
      #   @param blk [Proc] A block that describes the contents of the form
      #
      # @return [Hanami::Helpers::FormHelper::FormBuilder] the form builder
      #
      # @since 0.2.0
      #
      # @see Hanami::Helpers::FormHelper
      # @see Hanami::Helpers::FormHelper::Form
      # @see Hanami::Helpers::FormHelper::FormBuilder
      #
      # @example Inline Values In Template
      #   <%=
      #     form_for :book, routes.books_path, class: 'form-horizontal' do
      #       div do
      #         label      :title
      #         text_field :title, class: 'form-control'
      #       end
      #
      #       submit 'Create'
      #     end
      #   %>
      #
      #   <!-- output -->
      #
      #   <form action="/books" method="POST" accept-charset="utf-8" id="book-form" class="form-horizontal">
      #     <input type="hidden" name="_csrf_token" value="920cd5bfaecc6e58368950e790f2f7b4e5561eeeab230aa1b7de1b1f40ea7d5d">
      #     <div>
      #       <label for="book-title">Title</label>
      #       <input type="text" name="book[title]" id="book-title" value="Test Driven Development">
      #     </div>
      #
      #     <button type="submit">Create</button>
      #   </form>
      #
      #
      #
      # @example Use In A View
      #
      #   module Web::Views::Books
      #     class New
      #
      #     def form
      #       form_for :book, routes.books_path, class: 'form-horizontal' do
      #         div do
      #           label      :title
      #           text_field :title, class: 'form-control'
      #         end
      #
      #         submit 'Create'
      #       end
      #     end
      #   end
      #
      #   <!-- in the corresponding template use this -->
      #   <%= form %>
      #
      #   <!-- output -->
      #
      #   <form action="/books" method="POST" accept-charset="utf-8" id="book-form" class="form-horizontal">
      #     <input type="hidden" name="_csrf_token" value="920cd5bfaecc6e58368950e790f2f7b4e5561eeeab230aa1b7de1b1f40ea7d5d">
      #     <div>
      #       <label for="book-title">Title</label>
      #       <input type="text" name="book[title]" id="book-title" value="Test Driven Development">
      #     </div>
      #
      #     <button type="submit">Create</button>
      #   </form>
      #
      # @example Share Code Between Views
      #
      #   # Given the following views to create and update a resource
      #   module Web::Views::Books
      #     class New
      #       include Web::View
      #
      #       def form
      #         Form.new(:book, routes.books_path)
      #       end
      #
      #       def submit_label
      #         'Create'
      #       end
      #     end
      #
      #     class Edit
      #       include Web::View
      #
      #       def form
      #         Form.new(:book, routes.book_path(id: book.id),
      #           {book: book}, {method: :patch})
      #       end
      #
      #       def submit_label
      #         'Update'
      #       end
      #     end
      #   end
      #
      #   # The respective templates can be identical:
      #
      #   ## books/new.html.erb
      #   <%= render partial: 'books/form' %>
      #
      #   ## books/edit.html.erb
      #   <%= render partial: 'books/form' %>
      #
      #   # While the partial can have the following markup:
      #
      #   ## books/_form.html.erb
      #   <%=
      #     form_for form, class: 'form-horizontal' do
      #       div do
      #         label      :title
      #         text_field :title, class: 'form-control'
      #       end
      #
      #       submit submit_label
      #     end
      #   %>
      #
      #   <!-- output -->
      #
      #   <form action="/books" method="POST" accept-charset="utf-8" id="book-form" class="form-horizontal">
      #     <input type="hidden" name="_csrf_token" value="920cd5bfaecc6e58368950e790f2f7b4e5561eeeab230aa1b7de1b1f40ea7d5d">
      #     <div>
      #       <label for="book-title">Title</label>
      #       <input type="text" name="book[title]" id="book-title" value="Test Driven Development">
      #     </div>
      #
      #     <button type="submit">Create</button>
      #   </form>
      #
      # @example Method override
      #   <%=
      #     form_for :book, routes.book_path(id: book.id), method: :put do
      #       text_field :title
      #
      #       submit 'Update'
      #     end
      #   %>
      #
      #   <!-- output -->
      #
      #   <form action="/books/23" accept-charset="utf-8" id="book-form" method="POST">
      #     <input type="hidden" name="_method" value="PUT">
      #     <input type="hidden" name="_csrf_token" value="920cd5bfaecc6e58368950e790f2f7b4e5561eeeab230aa1b7de1b1f40ea7d5d">
      #     <input type="text" name="book[title]" id="book-title" value="Test Driven Development">
      #
      #     <button type="submit">Update</button>
      #   </form>
      #
      # @example Nested fields
      #   <%=
      #     form_for :delivery, routes.deliveries_path do
      #       text_field :customer_name
      #
      #       fields_for :address do
      #         text_field :city
      #       end
      #
      #       submit 'Create'
      #     end
      #   %>
      #
      #   <!-- output -->
      #
      #   <form action="/deliveries" accept-charset="utf-8" id="delivery-form" method="POST">
      #     <input type="hidden" name="_csrf_token" value="920cd5bfaecc6e58368950e790f2f7b4e5561eeeab230aa1b7de1b1f40ea7d5d">
      #     <input type="text" name="delivery[customer_name]" id="delivery-customer-name" value="">
      #     <input type="text" name="delivery[address][city]" id="delivery-address-city" value="">
      #
      #     <button type="submit">Create</button>
      #   </form>
      #
      # @example Override params
      #   <%=
      #     form_for :song, routes.songs_path, params: { song: { title: "Envision" } } do
      #       text_field :title
      #
      #       submit 'Create'
      #     end
      #   %>
      #
      #   <!-- output -->
      #
      #   <form action="/songs" accept-charset="utf-8" id="song-form" method="POST">
      #     <input type="hidden" name="_csrf_token" value="920cd5bfaecc6e58368950e790f2f7b4e5561eeeab230aa1b7de1b1f40ea7d5d">
      #     <input type="text" name="song[title]" id="song-title" value="Envision">
      #
      #     <button type="submit">Create</button>
      #   </form>
      def form_for(url, **attributes, &blk)
        attributes[:action] = url

        form_values = attributes.delete(:values) || {} # FIXME: take values also from view's exposures
        form_params = attributes.delete(:params) || params || {}
        values = Values.new(form_values, form_params, csrf_token)

        FormBuilder.new(values: values, **attributes, &blk).to_s
      end

      # Returns CSRF Protection Token stored in session.
      #
      # It returns <tt>nil</tt> if sessions aren't enabled or the value is missing.
      #
      # @return [String,NilClass] token, if present
      #
      # @since 0.2.0
      def csrf_token
        if defined?(session)
          session[CSRF_TOKEN]
        elsif defined?(locals) && locals[:session]
          locals[:session][CSRF_TOKEN]
        end
      end

      # Prints CSRF meta tags for Unobtrusive JavaScript (UJS) purposes.
      #
      # @return [Hanami::Helpers::HtmlHelper::HtmlBuilder,NilClass] the tags if `csrf_token` is not `nil`
      #
      # @since 1.2.0
      #
      # @example
      #   <html>
      #     <head>
      #       <!-- ... -->
      #       <%= csrf_meta_tags %>
      #     </head>
      #     <!-- ... -->
      #   </html>
      #
      #   <html>
      #     <head>
      #       <!-- ... -->
      #       <meta name="csrf-param" content="_csrf_token">
      #       <meta name="csrf-token" content="4a038be85b7603c406dcbfad4b9cdf91ec6ca138ed6441163a07bb0fdfbe25b5">
      #     </head>
      #     <!-- ... -->
      #   </html>
      def csrf_meta_tags
        return if csrf_token.nil?

        html.meta(name: "csrf-param", content: CSRF_TOKEN) +
          html.meta(name: "csrf-token", content: csrf_token)
      end
    end
  end
end
